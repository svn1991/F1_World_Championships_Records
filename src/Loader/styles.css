#loading-wrapper {
  width: 100%;
  height: 125px;
  background: white;
  position: relative;
}

#loading-text {
  display: block;
  position: absolute;
  top: 50%;
  left: 50%;
  color: #999;
  width: 100px;
  height: 30px;
  margin: -18px 0 0 -48px;
  text-align: center;
  font-family: "PT Sans Narrow", sans-serif;
  font-size: 10px;
}

#loading-content {
  display: block;
  position: relative;
  width: 100px;
  height: 100px;
  margin: 15px auto;
}

#loading-content:after {
  content: "";
  position: absolute;
  border: 3px solid #0f0;
  left: 15px;
  right: 15px;
  top: 15px;
  bottom: 15px;
}

#loading-content:before {
  content: "";
  position: absolute;
  border: 3px solid #00f;
  left: 5px;
  right: 5px;
  top: 5px;
  bottom: 5px;
}

#loading-content {
  border: 3px solid transparent;
  border-top-color: #4d658d;
  border-bottom-color: #4d658d;
  border-radius: 50%;
  -webkit-animation: loader 2s linear infinite;
  -moz-animation: loader 2s linear infinite;
  -o-animation: loader 2s linear infinite;
  animation: loader 2s linear infinite;
}

#loading-content:before {
  border: 3px solid transparent;
  border-top-color: #d4cc6a;
  border-bottom-color: #d4cc6a;
  border-radius: 50%;
  -webkit-animation: loader 3s linear infinite;
  -moz-animation: loader 2s linear infinite;
  -o-animation: loader 2s linear infinite;
  animation: loader 3s linear infinite;
}

#loading-content:after {
  border: 3px solid transparent;
  border-top-color: #84417c;
  border-bottom-color: #84417c;
  border-radius: 50%;
  -webkit-animation: loader 1.5s linear infinite;
  animation: loader 1.5s linear infinite;
  -moz-animation: loader 2s linear infinite;
  -o-animation: loader 2s linear infinite;
}

@-webkit-keyframes loaders {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}

@keyframes loader {
  0% {
    -webkit-transform: rotate(0deg);
    -ms-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    -ms-transform: rotate(360deg);
    transform: rotate(360deg);
  }
}
