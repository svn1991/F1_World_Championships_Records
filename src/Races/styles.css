.races-wrapper {
  width: 80%;
  margin: 20px auto;
}

.races-wrapper .races-title {
  margin: 0 0 20px 10px;
  font-size: 30px;
  font-weight: 100;
}

.races-wrapper #loading-wrapper {
  margin: 30px 0 !important;
}

.races-wrapper .react-bootstrap-table .table thead th {
  border-top: none;
}

.races-wrapper .react-bootstrap-table .highlight-winner {
  background-color: aquamarine;
  padding: 1px 10px;
}
