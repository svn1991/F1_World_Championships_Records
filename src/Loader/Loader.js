import React from 'react';
import './styles.css';

const Loader = () => (
  <div id="loading-wrapper">
    <div id="loading-text">LOADING</div>
    <div id="loading-content" />
  </div>
);

export default Loader;
