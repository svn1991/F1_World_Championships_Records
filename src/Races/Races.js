import React, { useState, useEffect } from 'react';
import BootstrapTable from 'react-bootstrap-table-next';
import PropTypes from 'prop-types';
import { getRacesInfoFrom } from '../CacheData/CacheData';

import Loader from '../Loader/Loader';

import './styles.css';

function Races(props) {
  const { year, seasonWinner } = props;
  const [races, setRaces] = useState({});

  useEffect(() => {
    getRacesInfoFrom(year).then(racesResult => {
      const raceObject = {};
      racesResult.forEach(race => {
        raceObject[race.round] = {
          round: race.round,
          raceName: race.raceName,
          raceUrl: race.url,
          winner: race.Results[0].Driver,
        };
      });
      setRaces(raceObject);
    });
  }, [year]);

  function getRacesProducts() {
    const raceKeys = Object.keys(races);
    return raceKeys.map(raceKey => {
      const race = races[raceKey];
      return {
        id: race.round,
        race: <a href={race.raceUrl}>{race.raceName}</a>,
        winner: (
          <a
            href={race.winner.url}
            className={
              race.winner.driverId === seasonWinner ? 'highlight-winner' : ''
            }
          >
            {`${race.winner.givenName} ${race.winner.familyName}`}
          </a>
        ),
      };
    });
  }

  const columns = [
    {
      dataField: 'race',
      text: 'Races',
    },
    {
      dataField: 'winner',
      text: 'Winner',
    },
  ];

  return (
    <div className="races-wrapper">
      <h1 className="races-title">{`F1 Races for Season ${year}`}</h1>
      {Object.keys(races).length === 0 ? (
        <Loader />
      ) : (
        <div data-testid="table">
          <BootstrapTable
            keyField="id"
            data={getRacesProducts()}
            columns={columns}
            condensed
            bordered={false}
          />
        </div>
      )}
    </div>
  );
}

Races.propTypes = {
  year: PropTypes.number.isRequired,
  seasonWinner: PropTypes.string.isRequired,
};

export default Races;
