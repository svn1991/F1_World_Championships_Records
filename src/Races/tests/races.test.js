import React from 'react';
import { render, cleanup } from '@testing-library/react';
import Races from '../Races';
import Loader from '../../Loader/Loader';
import * as RacesData from './races-2009.json';

const racesInfo = RacesData.MRData.RaceTable.Races;

const window = global;
const storage = window.localStorage;

describe('<Races />', () => {
  afterEach(() => {
    storage.clear();
    cleanup();
  });

  it.skip('Expect to log errors in console', done => {
    const spy = jest.spyOn(global.console, 'error');
    render(<Races />);
    requestAnimationFrame(() => {
      expect(spy).toHaveBeenCalled();
      done();
    });
  });

  // FIXME: correct this test, act render is having an issue
  it.skip('Expect to not log errors in console', done => {
    storage.setItem('https://ergast.com/api/f1/2009/results/1.json', JSON.stringify([]));
    const spy = jest.spyOn(global.console, 'error');
    render(<Races year={2009} seasonWinner="button" />);
    requestAnimationFrame(() => {
      expect(spy).not.toHaveBeenCalled();
      done();
    });
  });

  it('should render and match the snapshot', done => {
    storage.setItem('https://ergast.com/api/f1/2009/results/1.json', JSON.stringify([]));
    const { container: { firstChild } } = render(<Races year={2009} seasonWinner="button" />);
    requestAnimationFrame(() => {
      expect(firstChild).toMatchSnapshot();
      done();
    });
  });

  it('should render and match the snapshot with races data', done => {
    storage.setItem('https://ergast.com/api/f1/2009/results/1.json', JSON.stringify(racesInfo));
    const { container } = render(<Races year={2009} seasonWinner="button" />);

    requestAnimationFrame(() => {
      expect(container.firstChild).toMatchSnapshot();
      done();
    });
  });

  it('should mount loader component when no race information is available', done => {
    storage.setItem('https://ergast.com/api/f1/2009/results/1.json', JSON.stringify([]));
    const { container } = render(<Races year={2009} seasonWinner="button" />);
    const loader = render(<Loader />);

    requestAnimationFrame(() => {
      const raceLoading = container.querySelector('.races-wrapper #loading-wrapper');
      const loadContainer = loader.container.querySelector('#loading-wrapper');
      expect(raceLoading).toEqual(loadContainer);
      done();
    });
  });

  it('check driver who has won the season races vs driver who has only won races and not season', done => {
    storage.setItem('https://ergast.com/api/f1/2009/results/1.json', JSON.stringify(racesInfo));
    const { container } = render(<Races year={2009} seasonWinner="button" />);

    requestAnimationFrame(() => {
      const tableBodyRows = container.querySelectorAll('table tbody tr');

      const seasonWinner = tableBodyRows[3];
      expect(seasonWinner).toEqual(expect.anything());
      let columns = seasonWinner.children;
      expect(columns[0].children[0].innerHTML).toEqual('Bahrain Grand Prix');
      expect(columns[0].children[0].href).toEqual('http://en.wikipedia.org/wiki/2009_Bahrain_Grand_Prix');

      expect(columns[1].children[0].innerHTML).toEqual('Jenson Button');
      expect(columns[1].children[0].href).toEqual('http://en.wikipedia.org/wiki/Jenson_Button');
      expect(columns[1].children[0].classList.contains('highlight-winner')).toEqual(true);

      const raceWinner = tableBodyRows[8];
      expect(raceWinner).toEqual(expect.anything());
      columns = raceWinner.children;
      expect(columns[0].children[0].innerHTML).toEqual('German Grand Prix');
      expect(columns[0].children[0].href).toEqual('http://en.wikipedia.org/wiki/2009_German_Grand_Prix');

      expect(columns[1].children[0].innerHTML).toEqual('Mark Webber');
      expect(columns[1].children[0].href).toEqual('http://en.wikipedia.org/wiki/Mark_Webber');
      expect(columns[1].children[0].classList.contains('highlight-winner')).toEqual(false);
      done();
    });
  });
});
