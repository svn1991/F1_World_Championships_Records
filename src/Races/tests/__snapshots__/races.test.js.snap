// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Races /> should render and match the snapshot 1`] = `
<div
  class="races-wrapper"
>
  <h1
    class="races-title"
  >
    F1 Races for Season 2009
  </h1>
  <div
    id="loading-wrapper"
  >
    <div
      id="loading-text"
    >
      LOADING
    </div>
    <div
      id="loading-content"
    />
  </div>
</div>
`;

exports[`<Races /> should render and match the snapshot with races data 1`] = `
<div
  class="races-wrapper"
>
  <h1
    class="races-title"
  >
    F1 Races for Season 2009
  </h1>
  <div
    data-testid="table"
  >
    <div
      class="react-bootstrap-table"
    >
      <table
        class="table table-condensed"
      >
        <thead>
          <tr>
            <th
              tabindex="0"
            >
              Races
            </th>
            <th
              tabindex="0"
            >
              Winner
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Australian_Grand_Prix"
              >
                Australian Grand Prix
              </a>
            </td>
            <td>
              <a
                class="highlight-winner"
                href="http://en.wikipedia.org/wiki/Jenson_Button"
              >
                Jenson Button
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Malaysian_Grand_Prix"
              >
                Malaysian Grand Prix
              </a>
            </td>
            <td>
              <a
                class="highlight-winner"
                href="http://en.wikipedia.org/wiki/Jenson_Button"
              >
                Jenson Button
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Chinese_Grand_Prix"
              >
                Chinese Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Sebastian_Vettel"
              >
                Sebastian Vettel
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Bahrain_Grand_Prix"
              >
                Bahrain Grand Prix
              </a>
            </td>
            <td>
              <a
                class="highlight-winner"
                href="http://en.wikipedia.org/wiki/Jenson_Button"
              >
                Jenson Button
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Spanish_Grand_Prix"
              >
                Spanish Grand Prix
              </a>
            </td>
            <td>
              <a
                class="highlight-winner"
                href="http://en.wikipedia.org/wiki/Jenson_Button"
              >
                Jenson Button
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Monaco_Grand_Prix"
              >
                Monaco Grand Prix
              </a>
            </td>
            <td>
              <a
                class="highlight-winner"
                href="http://en.wikipedia.org/wiki/Jenson_Button"
              >
                Jenson Button
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Turkish_Grand_Prix"
              >
                Turkish Grand Prix
              </a>
            </td>
            <td>
              <a
                class="highlight-winner"
                href="http://en.wikipedia.org/wiki/Jenson_Button"
              >
                Jenson Button
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_British_Grand_Prix"
              >
                British Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Sebastian_Vettel"
              >
                Sebastian Vettel
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_German_Grand_Prix"
              >
                German Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Mark_Webber"
              >
                Mark Webber
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Hungarian_Grand_Prix"
              >
                Hungarian Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Lewis_Hamilton"
              >
                Lewis Hamilton
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_European_Grand_Prix"
              >
                European Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Rubens_Barrichello"
              >
                Rubens Barrichello
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Belgian_Grand_Prix"
              >
                Belgian Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Kimi_R%C3%A4ikk%C3%B6nen"
              >
                Kimi Räikkönen
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Italian_Grand_Prix"
              >
                Italian Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Rubens_Barrichello"
              >
                Rubens Barrichello
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Singapore_Grand_Prix"
              >
                Singapore Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Lewis_Hamilton"
              >
                Lewis Hamilton
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Japanese_Grand_Prix"
              >
                Japanese Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Sebastian_Vettel"
              >
                Sebastian Vettel
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Brazilian_Grand_Prix"
              >
                Brazilian Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Mark_Webber"
              >
                Mark Webber
              </a>
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="http://en.wikipedia.org/wiki/2009_Abu_Dhabi_Grand_Prix"
              >
                Abu Dhabi Grand Prix
              </a>
            </td>
            <td>
              <a
                class=""
                href="http://en.wikipedia.org/wiki/Sebastian_Vettel"
              >
                Sebastian Vettel
              </a>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
